### PR TITLE
Added UI support for final

### PR DESCRIPTION
**Pull request by KateMengjunXia**
https://github.com/nanoPayinc/NANOPAY/issues/5290

### CODE REVIEW

**Review by jlhughes — APPROVED** [review]

File: src/foam/core/Property.js [hunk 1]
```diff
@@ -534,8 +534,10 @@ foam.CLASS({
           // If this is the result of a factory setting the initial value,
           // then don't fire a property change event, since it hasn't
           // really changed.
-          if ( ! factory || oldValue !== undefined )
+          if ( ! factory || oldValue !== undefined ) {
+            if ( prop.final ) prop.visibility = 'RO';
```
Comment (TW80000) [inline src/foam/core/Property.js]: We have an `if ( isFinal ) { ... }` block a few lines above, wouldn't this make more sense there?

Comment (KateMengjunXia) [reply src/foam/core/Property.js]: I agree. Just updated the code.

**Review by kgrgreer — CHANGES_REQUESTED** [review]
I disagree with this PR.
If a property has final: true that means that you can only set the value once, and then it becomes a constant.
visibility: FINAL means that the value can only be edited in create mode, but becomes RO in edit mode. This should be the defaault for primaryKeys or values like createdBy or createdAt. A final: true property wouldn't be editable in the GUI because the value would get stuck on your first proposed value.

**Comment by TW80000** [comment]
> A final: true property wouldn't be editable in the GUI because the value would get stuck on your first proposed value.

It seems to me like it would make more sense if all `final: true` properties automatically were `visibility: FINAL` as well since they seem to be the same thing in spirit. If our view implementation is setting the property before the form is submitted, then isn't that a problem? When would you want a `final: true` property that wasn't `visibility: FINAL`? @kgrgreer

**Comment by kgrgreer** [comment]
Not the same thing. Say I'm in a GUI and I'm editing a unit test. It has a unique name property which is the primary key and is visibility: final. I might call it test1, then think that isn't a good name and rename it to 'CreatedByAware test', and then rename it finally to 'CreatedByAware creation test'. In this case I update the name property three times, or in on key mode was true then I updated it for each character I typed. If the property were final: true the value would lock up after I set it to test1, or even to just 't' in on key mode. As you're updating the GUI the property value keeps getting updated, but this would be prevented with final: true. final: is final once initially set to any value, visibility: final is final once the object is stored in the DAO. The first one is just used for a few internal high-performance use-cases and isn't at all useful for GUI's.  Java has 'final' variables which is what this feature is emulating.

**Comment by TW80000** [comment]
@kgrgreer Thanks for the explanation. I was arguing that we could change the way our inputs works to not actually set the values for the properties they're there for and leave the actual setting of the property to the form submission. That way the values would only be set once and therefore we could make all `final: true` properties `visibility: FINAL` with no problems, but after thinking about it more I disagree with that argument anyway.

**Pull request closed without merging**